### PR TITLE
Image Transport rework port in progress for review

### PR DESCRIPTION
**Pull request by mjcarroll**
Builds upon: ros2/message_filters#5

Port of image_transport to work with ROS2.

To use, the branch from the ros2/message_filters repo above needs to be used.

There's now more commits in the history than I plan to have at the end. Several of the later ones are reverting non-essential changes earlier. I plan to find where they were changed and squash the revert of the changes down.

### CODE REVIEW

**Comment by j-rivero** [comment]
Small patch to export headers and libraries to downstream package and the ament_package() to enable the full support inside ament (install setup.* files and some other friends):
```diff
diff --git a/image_transport/CMakeLists.txt b/image_transport/CMakeLists.txt
index e19da36..4c7fbe3 100644
--- a/image_transport/CMakeLists.txt
+++ b/image_transport/CMakeLists.txt
@@ -59,6 +59,9 @@ install(
   DESTINATION include
 )
 
+ament_export_include_directories(include)
+ament_export_libraries(${PROJECT_NAME})
+
 if(BUILD_TESTING)
   find_package(ament_lint_auto REQUIRED)
   ament_lint_auto_find_test_dependencies()
@@ -90,3 +93,5 @@ if(BUILD_TESTING)
     target_link_libraries(${PROJECT_NAME}-single_subscriber_publisher ${PROJECT_NAME})
   endif()
 endif()
+
+ament_package()
```

After these changes (and source install/setup.bash) we have one test failing:

```diff
4: Test timeout computed to be: 60
4: -- run_test.py: invoking following command in '/home/jrivero/code/ros2/image_transport_ws/src/image_common/image_transport':
4:  - /home/jrivero/code/ros2/image_transport_ws/build/image_transport/image_transport-message_passing --gtest_output=xml:/home/jrivero/code/ros2/image_transport_ws/build/image_transport/test_results/image_transport/image_transport-message_passing.gtest.xml
4: [==========] Running 2 tests from 1 test case.
4: [----------] Global test environment set-up.
4: [----------] 2 tests from MessagePassingTesting
4: [ RUN      ] MessagePassingTesting.one_message_passing
4: [WARN] [pluginlib.ClassLoader]: given plugin name 'lib/libimage_transport_plugins' should be '/libimage_transport_plugins' for better portability
4: unknown file: Failure
4: C++ exception with description "No plugins found! Does `rospack plugins --attrib=plugin image_transport` find any packages?" thrown in the test body.
4: [  FAILED  ] MessagePassingTesting.one_message_passing (18 ms)
4: [ RUN      ] MessagePassingTesting.stress_message_passing
4: [WARN] [pluginlib.ClassLoader]: given plugin name 'lib/libimage_transport_plugins' should be '/libimage_transport_plugins' for better portability
4: unknown file: Failure
4: C++ exception with description "No plugins found! Does `rospack plugins --attrib=plugin image_transport` find any packages?" thrown in the test body.
4: [  FAILED  ] MessagePassingTesting.stress_message_passing (13 ms)
4: [----------] 2 tests from MessagePassingTesting (31 ms total)
4: 
4: [----------] Global test environment tear-down
4: [==========] 2 tests from 1 test case ran. (31 ms total)
4: [  PASSED  ] 0 tests.
4: [  FAILED  ] 2 tests, listed below:
4: [  FAILED  ] MessagePassingTesting.one_message_passing
4: [  FAILED  ] MessagePassingTesting.stress_message_passing
```

And a bunch of QA linters are also complaining:
```
40% tests passed, 6 tests failed out of 10

Label Time Summary:
copyright     =   0.43 sec*proc (1 test)
cppcheck      =   0.53 sec*proc (1 test)
cpplint       =   1.77 sec*proc (1 test)
gtest         =   0.71 sec*proc (5 tests)
lint_cmake    =   0.40 sec*proc (1 test)
linter        =   3.77 sec*proc (5 tests)
uncrustify    =   0.64 sec*proc (1 test)

Total Test time (real) =   4.48 sec

The following tests FAILED:
	  4 - image_transport-message_passing (Failed)
	  6 - copyright (Failed)
	  7 - cppcheck (Failed)
	  8 - cpplint (Failed)
	  9 - lint_cmake (Failed)
	 10 - uncrustify (Failed)
Errors while running CTest
```
I can help fixing some of these linters. Let me know which ones do you want me to take care of.

**Comment by j-rivero** [comment]
> After these changes (and source install/setup.bash) we have one test failing:

The latest changes make the tests to pass for me.
```
The following tests passed:
	image_transport-camera_common
	image_transport-publisher
	image_transport-subscriber
	image_transport-message_passing
	image_transport-single_subscriber_publisher

100% tests passed, 0 tests failed out of 5
```
To fix the lint_cmake, a minimal patch is required:
```diff
diff --git a/image_transport/tutorial/CMakeLists.txt b/image_transport/tutorial/CMakeLists.txt
index ed5f28f..6b8455a 100644
--- a/image_transport/tutorial/CMakeLists.txt
+++ b/image_transport/tutorial/CMakeLists.txt
@@ -5,7 +5,7 @@ find_package(catkin REQUIRED COMPONENTS cv_bridge image_transport message_genera
 
 # add the resized image message
 add_message_files(DIRECTORY msg
-   FILES ResizedImage.msg
+  FILES ResizedImage.msg
 )
 generate_messages(DEPENDENCIES sensor_msgs)
```
To fix the other linters a bit more of work is needed. We can disable them by now or spend some hours fixing them.

File: image_transport/include/image_transport/simple_subscriber_plugin.h
Comment (tfoote) [inline image_transport/include/image_transport/simple_subscriber_plugin.h]: Add a TODO to implement this

File: image_transport/include/image_transport/simple_subscriber_plugin.h
Comment (tfoote) [inline image_transport/include/image_transport/simple_subscriber_plugin.h]: This needs a TODO too.

File: image_transport/include/image_transport/subscriber_filter.h
Comment (tfoote) [inline image_transport/include/image_transport/subscriber_filter.h]: missing space

File: image_transport/include/image_transport/subscriber_plugin.h
Comment (tfoote) [inline image_transport/include/image_transport/subscriber_plugin.h]: Is there a reason to cut the shared_ptr implementation/option here?

File: image_transport/src/camera_publisher.cpp
Comment (tfoote) [inline image_transport/src/camera_publisher.cpp:74]: A todo to restore support for these callbacks would be good. Maybe they shouldn't be in the constructor, but as an additional method etc.

File: image_transport/src/camera_publisher.cpp
Comment (tfoote) [inline image_transport/src/camera_publisher.cpp]: Needs TODO

File: image_transport/src/camera_publisher.cpp
Comment (tfoote) [inline image_transport/src/camera_publisher.cpp]: TODO or use https://github.com/ros2/rcutils/pull/112 after merge

File: image_transport/src/camera_publisher.cpp
Comment (tfoote) [inline image_transport/src/camera_publisher.cpp]: TODO or use ros2/rcutils#112 after merge

File: image_transport/src/camera_subscriber.cpp
Comment (tfoote) [inline image_transport/src/camera_subscriber.cpp]: reset the pointers here?

File: image_transport/src/camera_subscriber.cpp
Comment (tfoote) [inline image_transport/src/camera_subscriber.cpp]: Can use RCUTILS_LOG_WARN

File: image_transport/src/camera_subscriber.cpp
Comment (tfoote) [inline image_transport/src/camera_subscriber.cpp]: Add TODO

File: image_transport/src/publisher.cpp
Comment (tfoote) [inline image_transport/src/publisher.cpp]: TODO resolved with local implementation

File: image_transport/src/publisher.cpp
Comment (tfoote) [inline image_transport/src/publisher.cpp]: More places here for debug macros instead of fprintf

**Review by tfoote — COMMENTED** [review]
Overall this looks good. There are a few places to improve the debug outputs from fprintf and also a few places to add TODOs where things were skipped for the MVP

File: image_transport/src/camera_subscriber.cpp
Comment (mjcarroll) [reply image_transport/src/camera_subscriber.cpp]: Which pointers?  I think that the only one in this scope is the timer pointer?

**Comment by mjcarroll** [comment]
* Linux [Build Status](http://ci.ros2.org/job/ci_linux/5017/)
* Linux-aarch64 [Build Status](http://ci.ros2.org/job/ci_linux-aarch64/1827/)
* macOS [Build Status](http://ci.ros2.org/job/ci_osx/4173/)
* Windows [Build Status](http://ci.ros2.org/job/ci_windows/5028/)

File: image_transport/src/camera_subscriber.cpp
Comment (tfoote) [reply image_transport/src/camera_subscriber.cpp]: The `info_sub_` and `image_sub_` are shared pointers to the subscribers. If we're not explicitly calling unsubscribe on them they'll keep subscribing to data. So to get them to stop subscribing you can reset the shared pointers `info_sub_.reset()` to stop them from subscribing, assuming that no one else has stored a reference to them.

File: image_transport/include/image_transport/subscriber_plugin.h
Comment (j-rivero) [inline image_transport/include/image_transport/subscriber_plugin.h:38]: while implementing the theora plugin I made the following changes to be able to compile the package:

```diff
diff --git a/image_transport/include/image_transport/subscriber_plugin.h b/image_transport/include/image_transport/subscriber_plugin.h
index 61d5011..0f0afb0 100644
--- a/image_transport/include/image_transport/subscriber_plugin.h
+++ b/image_transport/include/image_transport/subscriber_plugin.h
@@ -36,6 +36,8 @@
 #define IMAGE_TRANSPORT_SUBSCRIBER_PLUGIN_H
 
 #include <rclcpp/macros.hpp>
+#include <rclcpp/node.hpp>
+#include <sensor_msgs/msg/image.hpp>
 
 namespace image_transport
 {
```

File: image_transport/CMakeLists.txt
Comment (mikaelarguedas) [inline image_transport/CMakeLists.txt]: is `tinyxml_vendor`/`tinyxml` used anywhere in this package?

File: image_transport/CMakeLists.txt
Comment (mikaelarguedas) [inline image_transport/CMakeLists.txt]: If there is a use case for this to be built statically (maybe there is not), we recommend using `ament_cmake_ros` to allow people to chose to build statically or dynamically via the `BUILD_SHARED_LIBS` CMake option

File: image_transport/CMakeLists.txt
Comment (mikaelarguedas) [inline image_transport/CMakeLists.txt]:
```
  ARCHIVE DESTINATION lib
  LIBRARY DESTINATION lib
  RUNTIME DESTINATION bin
```
For multi OS support in the future

File: image_transport/CMakeLists.txt
Comment (mikaelarguedas) [inline image_transport/CMakeLists.txt]: Executables should be installed to `DESTINATION lib/${PROJECT_NAME}` to be found by `ros2 run`

File: image_transport/CMakeLists.txt
Comment (mikaelarguedas) [inline image_transport/CMakeLists.txt]: Nit: quotes unnecessary here

File: image_transport/package.xml
Comment (mikaelarguedas) [inline image_transport/package.xml]: This dependency seems to be unused

File: image_transport/package.xml
Comment (mikaelarguedas) [inline image_transport/package.xml]: This dependency seems to be unused

File: image_transport/package.xml
Comment (mikaelarguedas) [inline image_transport/package.xml]: Nit: alphabetical order

File: image_transport/package.xml
Comment (mikaelarguedas) [inline image_transport/package.xml:1]: We should update maintainer names to avoid original maintainers receiving email when we start testing / releasing this on the ROS 2 farm

**Review by mikaelarguedas — COMMENTED** [review]
a quick review just for the CMake and package.xml part of the PR

File: image_transport/include/image_transport/subscriber_plugin.h
Comment (mjcarroll) [reply image_transport/include/image_transport/subscriber_plugin.h:38]: Included, thanks.

File: image_transport/package.xml
Comment (mjcarroll) [reply image_transport/package.xml:1]: Done.

File: image_transport/src/camera_publisher.cpp
Comment (mjcarroll) [reply image_transport/src/camera_publisher.cpp:74]: Added.

**Comment by j-rivero** [comment]
Changes looks good to me, it works well on my Bionic system. Great work Michael, thanks.

**Review by j-rivero — APPROVED** [review]

File: image_transport/include/image_transport/image_transport.h
Comment (dirk-thomas) [inline image_transport/include/image_transport/image_transport.h:118]: It would be helpful for porting existing code if *at least additionally) the previous function signature is maintained. So in this case keep a `queue_size` parameter and set it internally on the profile.

Same applies to other signatures.

Comment (dirk-thomas) [reply image_transport/include/image_transport/image_transport.h:118]: I guess there are several other API changes which will make it a not straight forward conversion at this point.

It looks like the hints have been removed from several signature but instead the transport string has been introduced. What is the plan around the `TransportHints`?

Comment (mjcarroll) [reply image_transport/include/image_transport/image_transport.h:118]: Can do for `queue_size`.
I had originally cut `TransportHints` because it would require a parameters_client internally in order to retrieve the string parameter.  The only other thing it was doing was interacting with the ROS1 transport hints, so it seemed easier to pass a string in from the outside.

I can make another similar structure if keeping the API the same is important.

Comment (dirk-thomas) [reply image_transport/include/image_transport/image_transport.h:118]: I have ported the `web_video_server` package already. I just noticed the effort during that work. I am not sure if anything should be done but every change in the API (as small as it might be) will be perceived as an (unnecessary?) burden by every user / developer when porting their code.

Comment (mjcarroll) [reply image_transport/include/image_transport/image_transport.h:118]: Sounds reasonable, can I do it in a second PR so I can get this one merged?

Comment (dirk-thomas) [reply image_transport/include/image_transport/image_transport.h:118]: Absolutely. Not a blocker.

**Comment by mjcarroll** [comment]
* Linux [Build Status](http://ci.ros2.org/job/ci_linux/5069/)
* Linux-aarch64 [Build Status](http://ci.ros2.org/job/ci_linux-aarch64/1862/)
* macOS [Build Status](http://ci.ros2.org/job/ci_osx/4209/)
* Windows [Build Status](http://ci.ros2.org/job/ci_windows/5060/)

**Comment by mjcarroll** [comment]
Windows and OSX failures expected, due to `cv_bridge` deps
Linux and aarch unstable due to failing `cv_bridge` tests.

**Comment by mikaelarguedas** [comment]
> Windows and OSX failures expected, due to cv_bridge deps

Can you clarify what deps are missing ?
If we expect this PR to be tested on all platforms, we should make sure CI can test it. Otherwise we can just "not run CI" for these PRs on Windows/MacOS

> Linux and aarch unstable due to failing cv_bridge tests.

While I don't think there is much we can do about the CMake warning. It would be valuable to modify the CI branch you use so that the `cv_bridge` tests pass so that we have a "clean" state (no failing tests).

**Comment by mjcarroll** [comment]
Looks like Boost on Windows, and boost_python3 on OS X, I'll see what I can do to get them fixed.

**Pull request merged**
